src/login.ts

```diff
@@ -0,0 +1,242 @@
+ * @returns - The JWT token.
+ */
+export async function login(
+  client: SkynetClient,
+  seed: Uint8Array,
+  email: string,
+  customOptions?: CustomLoginOptions
+): Promise<string> {
+  const opts = { ...DEFAULT_LOGIN_OPTIONS, ...client.customOptions, ...customOptions };
+
+  const { publicKey, privateKey } = genPortalLoginKeypair(seed, email);
+
+  const registerRequestResponse = await client.executeRequest({
+    endpointPath: opts.endpointLoginRequest,
+    method: "GET",
+    subdomain: "account",
+    query: { pubKey: publicKey },
+  });
+
+  const challenge = registerRequestResponse.data.challenge;
+  const portalRecipient = getPortalRecipient(await client.portalUrl());
+  const challengeResponse = signChallenge(privateKey, challenge, CHALLENGE_TYPE_LOGIN, portalRecipient);
+
+  const data = challengeResponse;
+  const registerResponse = await client.executeRequest({
+    endpointPath: opts.endpointLogin,
+    method: "POST",
+    subdomain: "account",
+    data,
+  });
+
+  // TODO: This is temporary for testing purposes. This should be changed to
+  // `set-cookie` (will only work in the browser).
+  const jwt = registerResponse.headers[COOKIE_HEADER_NAME];
+  return jwt;
+}
+
+/**
+ * Signs the given challenge.
+ *
+ * @param privateKey - The user's login private key.
+ * @param challenge - The challenge received from the server.
+ * @param challengeType - The type of the challenge.
+ * @param portalRecipient - The portal we are communicating with.
+ * @returns - The challenge response from the client.
+ */
+function signChallenge(
+  privateKey: string,
+  challenge: string,
+  challengeType: "skynet-portal-login" | "skynet-portal-register",
+  portalRecipient: string
+): ChallengeResponse {
+  validateHexString("challenge", challenge, "challenge from server");
+
+  const challengeBytes = hexToUint8Array(challenge);
+  validateUint8ArrayLen("challengeBytes", challengeBytes, "calculated challenge bytes", CHALLENGE_SIZE);
+
+  const typeBytes = stringToUint8ArrayUtf8(challengeType);
+
+  const portalBytes = stringToUint8ArrayUtf8(portalRecipient);
+
+  const dataBytes = new Uint8Array([...challengeBytes, ...typeBytes, ...portalBytes]);
+
+  const privateKeyBytes = hexToUint8Array(privateKey);
+  const signatureBytes = sign(dataBytes, privateKeyBytes).slice(0, CHALLENGE_SIGNATURE_SIZE);
+  validateUint8ArrayLen("signatureBytes", signatureBytes, "calculated signature", CHALLENGE_SIGNATURE_SIZE);
+
+  return {
+    response: toHexString(dataBytes),
+    signature: toHexString(signatureBytes),
+  };
+}
+
+/**
+ * Gets the portal recipient string from the portal URL, e.g. siasky.net =>
+ * siasky.net, dev1.siasky.dev => siasky.dev.
+ *
+ * @param portalUrl - The full portal URL.
+ * @returns - The shortened portal recipient name.
+ */
+function getPortalRecipient(portalUrl: string): string {
+  const url = new URL(portalUrl);
+
+  // Get last two portions of the hostname.
+  return url.hostname.split(".").slice(-2).join(".");
+}
```

Comment:
We are making a dangerous assumption about the portal's name. It might very well be a portal operating at `skynet.mycompany.com` in which case the recipient will be rejected.

I understand that having this in place facilitates development and testing but there are safer ways to achieve that. We should be using the full portal name. We can test by locking onto a specific server by using our `/etc/hosts`.

Reply:
How do you suggest we get the portal name? Chris said we shouldn't trust anything the portal says about itself.

Reply:
That's correct. Do not rely on what the portal is telling you about itself. Rely on the request you are making to it (URL). This URL should be confirmed by a valid TLS certificate.

So, if the user is talking to `hello.portal.company.com` then that's the full name of the portal.

I am not sure how we should handle these cases:
* `skylink(base32/base64).hello.portal.company.com`
* `skapp.hns.hello.portal.company.com`

We might be able to use the server's SSL cert for this or we can mimic the resolutions we're doing in webportal. I really don't like the latter because it's brittle and it's easy to desync it.

Reply:
I specifically had in mind the case of `devX.siasky.dev`, where the portal name is `siasky.dev`. That's why I get the last two parts of the domain, for now.

In the SDK, we already do this in a safe way by getting the portal name from one of the headers -- but this falls under "trusting the portal".